### ISSUE TITLE
Log slow and unindexed queries (options)

### ISSUE THREAD

**Issue by zuiderkwast** [mysql-otp/mysql-otp#12]
In each result we get some status bits from the server with the meanings "query was slow", "no index was used", "no good index was used". When this happens, log the query using `error_logger:warning_msg("MySQL query was slow: ~s", [Query])`, etc.

Options:
- `{log_slow_queries, boolean()}`, default `false`
- <strike>`{log_no_index_used, boolean()}` default false? Performing queries with no index is always bad.</strike>
- <strike>`{log_no_good_index_used, boolean()}` default false?</strike>

I think we should log only slow queries and add the "index used" info to the "slow query" log entry. No index querys that are not slow are not logged. Examples:

```
"MySQL query (with no index) was slow: SELECT ..."
"MySQL query (with no good index) was slow: SELECT ..."
"MySQL query was slow: SELECT ..."
```

[Edited 20 March 2018]

**Comment by m13m** [comment]
Ping @zuiderkwast  can I take a shot. From the C/S protocol page 

Do we have to warn only for the following status flag?

`SERVER_STATUS_NO_GOOD_INDEX_USED`
`SERVER_STATUS_NO_INDEX_USED`
`SERVER_QUERY_WAS_SLOW`

Any suggestion before I start working on this?

**Comment by juhlig** [comment]
I wanted to give this a try, but really I could not find an (easy) way to cause the slow query flag to be set in the server status. I tried with MySQL 5.7 and `SELECT SLEEP(11)`, then got the connection process' state with `sys:get_state` to see what's in the status field, but there is always just a 2 (meaning autocommit). On the other hand, the query _does_ show up in the slow query log. Am I missing something here?

**Comment by zuiderkwast** [comment]
I don't know an easy way either. If it shows up in the slow query log but not in the client, I don't know what's happening. If there is a config option, I don't know about it.

The way I can think of is to populate the database with random data and no index or no good index, possibly with many workers to get some load on the database as well. There is no reason to run such a test by default on all OTP versions. A more sophisticated travis file would be necessary for that.

If we can reproduce it and capture the traffic (using tcpdump or ngrep), we can add it as a protocol unit test, i.e. mock the server (see `mysql_protocol_tests`). If we can't even make the server produce these warnings, we can write a server response by hand based on the protocol documentation.

**Comment by zuiderkwast** [comment]
Btw, it would be nice to use log using logger where available, with error_logger as a fallback, for this and the existing logging. This could be a separate issue.

**Comment by juhlig** [comment]
I give up on that for now =^^= I have been trying to get that flag to be set for a while yesterday, different MySQL versions, large tables, queries that ran well over 30s, adjusting the `long_query_time` server variable... all to no avail =^^= The `no_index` flag gets set all right, but never the `query_was_slow` one. So really, all we (I) can do is fake it for tests.

> Btw, it would be nice to use log using logger where available, with error_logger as a fallback, for this and the existing logging. This could be a separate issue.

In `ranch`, the `logger` can be set with a config option, defaulting to `logger` if not explicitly set. Would that be an option for you? Since we want to be compatible down to R16, the default would be `logger` in OTP versions that support it, and `error_logger` if not.

**Comment by juhlig** [comment]
Actually, that'd probably be overkill... So I'll just use logger and fall back to error_logger instead of making it configurable.

**Comment by zuiderkwast** [comment]
> The `no_index` flag gets set all right, but never the `query_was_slow` one. So really, all we (I) can do is fake it for tests.

If it's logged as a slow query on the server side, the bit should be set. If not, it ought to be a bug. Maybe this feature of MySQL was just documented but never working... Anyway, faking is not a good options, so I understand that you give it up for now.

> the logger can be set with a config option

Why an option? Why not just use logger when available and error_logger otherwise? Backwards compatibility concerns?

**Comment by juhlig** [comment]
> > The `no_index` flag gets set all right, but never the `query_was_slow` one. So really, all we (I) can do is fake it for tests.
> 
> If it's logged as a slow query on the server side, the bit should be set. If not, it ought to be a bug. Maybe this feature of MySQL was just documented but never working... Anyway, faking is not a good options, so I understand that you give it up for now.

Hm, not quite yet :) I just tried the same with ooold `erlmysql` (god, how complicated that one is...), it has a `queryWasSlow` detection built in as well, and what do you know, it works there XD I don't understand why, or rather why not in `mysql-otp`, I even went to the trouble of logging the very packets that were read from the server, and that bit was never set in there. Will investigate more tomorrow what the difference may be.

> > the logger can be set with a config option
> 
> Why an option? Why not just use logger when available and error_logger otherwise? Backwards compatibility concerns?

I reverted my opinion on that already, see above ;) I prepared that one already, but now I find it very difficult to get it working with that error logger capturing in the tests. Will see to that tomorrow, too.

**Comment by zuiderkwast** [comment]
Just a guess: They might actually check the clock to know if the query was slow.

Regarding logger: Yes, it's good if the log capturing in the tests can be updated to use logger whenever the implementation uses logger. Better make logger and slow query logging in two separate PRs.

**Comment by juhlig** [comment]
> Just a guess: They might actually check the clock to know if the query was slow.

No, checked that already, they take it from the status flags. But I'm closing in, almost there :)

> Regarding logger: Yes, it's good if the log capturing in the tests can be updated to use logger whenever the implementation uses logger.

Yeah, worked on that that already. Problem is that it's difficult to make the return from both variants "look" the same to the test suites.

> Better make logger and slow query logging in two separate PRs.

Sure.

**Comment by juhlig** [comment]
Ok, found it :) Apparently I overlooked something when I read the tracked packets from the server as well. You see, that slow query flag gets only set on the _final_ EOF packet from the server, _not_ on the EOF packet following the column definitions. But we just drop that final packet in `mysql_protocol:fetch_resultset_rows`, never examine it. That's why we never get this flag in the status in the state. I'll make an intermediate PR to fix this, before tackling slow query logging itself.

**Comment by zuiderkwast** [comment]
Solved in #150
